evm/src/cpu/columns.rs

```diff
@@ -1,207 +1,204 @@
+    pub mem_addr_virtual: [T; memory::NUM_CHANNELS],
+    pub mem_value: [[T; memory::VALUE_LIMBS]; memory::NUM_CHANNELS],
 }
 
-const MEM_ISREAD_START: usize = MEM_CHANNEL_USED_START + memory::NUM_CHANNELS;
-pub const fn mem_is_read(channel: usize) -> usize {
-    debug_assert!(channel < memory::NUM_CHANNELS);
-    MEM_ISREAD_START + channel
+// `u8` is guaranteed to have a `size_of` of 1.
+pub const NUM_CPU_COLUMNS: usize = size_of::<CpuColumnsView<u8>>();
+
+unsafe fn transmute_no_compile_time_size_checks<T, U>(value: T) -> U {
+    debug_assert_eq!(size_of::<T>(), size_of::<U>());
+    // Need ManuallyDrop so that `value` is not dropped by this function.
+    let value = ManuallyDrop::new(value);
+    // Copy the bit pattern. The original value is no longer safe to use.
+    transmute_copy(&value)
+}
+
+impl<T> From<[T; NUM_CPU_COLUMNS]> for CpuColumnsView<T> {
+    fn from(value: [T; NUM_CPU_COLUMNS]) -> Self {
+        unsafe { transmute_no_compile_time_size_checks(value) }
+    }
+}
+
+impl<T> From<CpuColumnsView<T>> for [T; NUM_CPU_COLUMNS] {
+    fn from(value: CpuColumnsView<T>) -> Self {
+        unsafe { transmute_no_compile_time_size_checks(value) }
+    }
+}
+
+impl<T> Borrow<CpuColumnsView<T>> for [T; NUM_CPU_COLUMNS] {
+    fn borrow(&self) -> &CpuColumnsView<T> {
+        unsafe { transmute(self) }
+    }
 }
 
-const MEM_ADDR_CONTEXT_START: usize = MEM_ISREAD_START + memory::NUM_CHANNELS;
-pub const fn mem_addr_context(channel: usize) -> usize {
-    debug_assert!(channel < memory::NUM_CHANNELS);
-    MEM_ADDR_CONTEXT_START + channel
+impl<T> BorrowMut<CpuColumnsView<T>> for [T; NUM_CPU_COLUMNS] {
+    fn borrow_mut(&mut self) -> &mut CpuColumnsView<T> {
+        unsafe { transmute(self) }
+    }
 }
 
-const MEM_ADDR_SEGMENT_START: usize = MEM_ADDR_CONTEXT_START + memory::NUM_CHANNELS;
-pub const fn mem_addr_segment(channel: usize) -> usize {
-    debug_assert!(channel < memory::NUM_CHANNELS);
-    MEM_ADDR_SEGMENT_START + channel
+impl<T> Deref for CpuColumnsView<T> {
+    type Target = [T; NUM_CPU_COLUMNS];
+    fn deref(&self) -> &Self::Target {
+        unsafe { transmute(self) }
+    }
 }
 
-const MEM_ADDR_VIRTUAL_START: usize = MEM_ADDR_SEGMENT_START + memory::NUM_CHANNELS;
-pub const fn mem_addr_virtual(channel: usize) -> usize {
-    debug_assert!(channel < memory::NUM_CHANNELS);
-    MEM_ADDR_VIRTUAL_START + channel
+impl<T> DerefMut for CpuColumnsView<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { transmute(self) }
+    }
 }
 
-const MEM_ADDR_VALUE_START: usize = MEM_ADDR_VIRTUAL_START + memory::NUM_CHANNELS;
-pub const fn mem_value(channel: usize, limb: usize) -> usize {
-    debug_assert!(channel < memory::NUM_CHANNELS);
-    debug_assert!(limb < memory::VALUE_LIMBS);
-    MEM_ADDR_VALUE_START + channel * memory::VALUE_LIMBS + limb
+const fn make_col_map() -> CpuColumnsView<usize> {
+    let mut indices_arr = [0; NUM_CPU_COLUMNS];
+    let mut i = 0;
```

Comment:
nit: `for i in 0..NUM_CPU_COLUMNS` seems more succinct, or could do `(0..NUM_CPU_COLUMNS).try_into().unwrap()`

Reply:
I don't think `for` loops are permitted in a `const fn` are they?

Reply:
Yeah, neither of those approaches is permitted in a `const` context

Reply:
Oh ok then 👍